internal/server/singleprocess/state/job.go

```diff
@@ -1382,6 +1400,57 @@ func (s *State) jobCandidateById(
 	return nil, nil
 }
 
+// jobCandidateByLabels returns the most promising candidate job to assign
+// that is targeting a specific runner by labels.
+func (s *State) jobCandidateByLabels(
+	memTxn *memdb.Txn, ws memdb.WatchSet, r *runnerIndex, assign bool,
+) (*jobIndex, error) {
+	iter, err := memTxn.LowerBound(
+		jobTableName,
+		jobQueueTimeIndexName,
+		pb.Job_QUEUED,
+		time.Unix(0, 0),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+
+		job := raw.(*jobIndex)
+		if job.State != pb.Job_QUEUED || job.TargetRunnerLabels == nil {
+			continue
+		}
+
+		// If this job is blocked, it is not a candidate.
+		if blocked, err := s.jobIsBlocked(memTxn, job, ws); err != nil {
+			return nil, err
+		} else if blocked && assign {
+			continue
+		}
+
+		// Check whether job target labels match with runner labels
+		match := true
+		for k, v := range job.TargetRunnerLabels {
+			if val, ok := r.Runner.Labels[k]; !ok || v != val {
+				match = false
```

Comment:
Wait, we would want to `break` out of the loop as soon as we hit a mismatch, and a continue outside of the loop to return `nil,nil` right? Otherwise a `continue` inside the loop would just go on to the next iteration?

Reply:
Oh you're right. Nevermind keep this. 😄